sign/dilithium/mode3/internal/kat_test.go

```diff
@@ -0,0 +1,118 @@
+package internal
+
+// Code to generate the NIST "PQCsignKAT" test vectors.
+// See PQCsignKAT_sign.c and randombytes.c in the reference implementation.
+
+import (
+	"crypto/aes"
+	"crypto/sha256"
+	"fmt"
+	"testing"
+)
+
+// From SHA256SUMS in the reference implementation
+var ExpectedHashes = map[string]string{
+	"Dilithium1":     "dd83f8584fded0398547827edff081969335c32069f3e4a9dbd865fd5c2ecd2b",
+	"Dilithium2":     "532f4a7a416bba96b607395a6d07fc0eaab1f1f968e49758d2a97c718de832e7",
+	"Dilithium3":     "37a16744627f2566180a547d022f03a36d22c50080303027179751070e626c72",
+	"Dilithium4":     "4c2e6d7c8675e9345e3ab7036a4e9fb786549d242462ba9b68f58db94e84147a",
+	"Dilithium1-AES": "68fabe91565c9a664d2461c7510ac32419eadfac0566dc3e9141d276bb98e11a",
+	"Dilithium2-AES": "08865a608edcdb5723769c583b37c17c9ff8cae578f1d88df7e173ed06dd23fa",
+	"Dilithium3-AES": "f3c5fcceafa9fb2462721f272791a26c9a123b3a07fad7e07dfec232085fdd7f",
+	"Dilithium4-AES": "8de4e2ac2032f714263aa0d045275ec62b6f192f8828cfe82b63ec0b0b32deb6",
+}
+
+// See NIST's PQCgenKAT.c.
+type DRBG struct {
+	key [32]byte
+	v   [16]byte
+}
+
+func (g *DRBG) incV() {
+	for j := 15; j >= 0; j-- {
+		if g.v[j] == 255 {
+			g.v[j] = 0
+		} else {
+			g.v[j]++
+			break
+		}
+	}
+}
+
+// AES256_CTR_DRBG_Update(pd, &g.key, &g.v)
+func (g *DRBG) update(pd *[48]byte) {
+	var buf [48]byte
+	b, _ := aes.NewCipher(g.key[:])
+	for i := 0; i < 3; i++ {
+		g.incV()
+		b.Encrypt(buf[i*16:(i+1)*16], g.v[:])
+	}
+	if pd != nil {
+		for i := 0; i < 48; i++ {
+			buf[i] ^= pd[i]
+		}
+	}
+	copy(g.key[:], buf[:32])
+	copy(g.v[:], buf[32:])
+}
+
+// randombyte_init(seed, NULL, 256)
+func NewDRBG(seed *[48]byte) (g DRBG) {
+	g.update(seed)
+	return
+}
+
+// randombytes
+func (g *DRBG) Fill(x []byte) {
+	var block [16]byte
+
+	b, _ := aes.NewCipher(g.key[:])
+	for len(x) > 0 {
+		g.incV()
+		b.Encrypt(block[:], g.v[:])
+		if len(x) < 16 {
+			copy(x[:], block[:len(x)])
+			break
+		}
+		copy(x[:], block[:])
+		x = x[16:]
+	}
+	g.update(nil)
+}
+
+func TestPQCgenKATSign(t *testing.T) {
```

Comment:
When (for some reason) this test fails, it would be difficult to know what was the expected output.

Reply:
If this fails, then it fails for, say, some signature.  If you just get the bad signature, you're still stuck.  Why is the signature bad?  So we'd have to add debug statements for the intermediate values, etc.  We'd need to dig down anyway then, so unfolding this first step already does not make a big difference in my opinion.

Reply:
However, as is, there are some other reasons for which the hash could not match.
For example, the KAT vector could differ in its format.
Also, it would be good to include the original test vector file in the /testdata folder.

Reply:
> For example, the KAT vector could differ in its format.

Of course, but that still would require some digging.  I would suspect the CI to truncate the difference between the expected values anyways.

> Also, it would be good to include the original test vector file in the /testdata folder.

The reference implementation does not include these test vectors — it also generates them on the fly.